Revamp makefile

tharvik opened the pull request:
## Changes
Makefile is gone out of sync since at least #1173 so I updated it (a bunch).
* rewrite it in a single, more standard makefile
* forward most commands to npm
* support parallel use of it (`make -j get_images` is nice to have)
* get rid of load-time auth check and trust user
* rework targets
  * `prep_commit` -> `format`, `lint`
  * `dev`, `beta`, `release` -> `build-dev`, `build-prod`, `build-tests`: match closer to npm
  * `prep_staging`, `package`: now uneeded
* update devguide accordingly
  * nicer split between vscode and CLI 
* ~~package.json: also remove unworking target~~ it was needed
* `npm run lint` now runs every linter
* bsconfig-tests create a pkg to have `make build-tests` behaving the same as the other build-* targets

## Fixes
#1401

cewert commented:
Thanks for the PR! Going to give the code a quick look and I'll give it a test when I can.

cewert commented on package.json:
```diff
@@ -27,7 +27,6 @@
     "build": "npx rimraf build/ out/ && npx bsc --project bsconfig.json",
     "build-prod": "npx rimraf build/ out/ && npx bsc --project bsconfig-prod.json",
     "build-tests": "npx rimraf build/ out/ && npx bsc --project bsconfig-tests.json",
-    "build-tdd": "npx rimraf build/ out/ && npx bsc --project bsconfig-tdd.json",
```
This line needs to be added back - It's not a dead target. It's meant to be used when writing unit tests that way you can run only the unit test you're working on instead of running and waiting on the whole suit of tests (like with `build-tests`). That's why the file was added to `.gitignore` - it's not meant to be committed to git. 

See https://github.com/jellyfin/jellyfin-roku/pull/1141 for more info

tharvik replied:
ho, I see, that's useful indeed, dropping commit then.

cewert commented on DEVGUIDE.md:
````diff
@@ -46,6 +46,7 @@ cd jellyfin-roku
 Install Dependencies:
 
 ```bash
+sudo apt-get install npm
````
I think it makes sense to add this somewhere but it doesn't really belong here IMO. It's under a header title `Clone the GitHub Repo` which doesn't really fit and it also doesn't help windows users.

Maybe we should make a new heading called `Install` in between clone the repo and Method 1? Then we could add npm install instructions for windows and linux/debian as well as moving the `npm install` command there. What do you think?

tharvik replied:
hum, good idea, we can even regroup most of the scattered dependencies instruction, there is a bunch of `apt-get` calls around.

for VSCode, it will use only `npm run` calls and take care of `npm install` for us, yes? so we even remove the `npm install` block as make already takes care of it if `node_modules` doesn't exists.

I've added a commit showing a possible regroupement, WDYT?

cewert replied:
> hum, good idea, we can even regroup most of the scattered dependencies instruction, there is a bunch of apt-get calls around.

I think it makes sense to group together the apt install commands but `make` and `curl` are only needed for method 3. So I think we need to specify that in the install section. npm on the other hand is needed by all 3 methods

> for VSCode, it will use only npm run calls and take care of npm install for us, yes? so we even remove the npm install block as make already takes care of it if node_modules doesn't exists.

When using vscode, devs still need to run `npm install`

cewert commented:
> arf, I force pushed the branch to remove the `package.json` commit, sorry about that: I wanted to avoid having a revert commit but didn't knew the policy there.

Good question. I don't think we have a policy on that. It's fine :)

> while at it, for `bsconfig-tests.json`, why does it actively doesn't create a package? running the suite via `make build-tests install` doesn't work as I wanted to

It's currently setup so that the vscode brightscript extension creates and deploys the testing app (the unit tests have to be ran on the actual device). You can change it to make a zip by setting the `createPackage` var to true [here](https://github.com/jellyfin/jellyfin-roku/blob/unstable/bsconfig-tests.json#L29).

cewert commented on DEVGUIDE.md:
````diff
@@ -83,19 +93,13 @@ Example:
 
 ```bash
-make dev
+make build-dev
````
```suggestion
npm run build
```

I think it makes more sense for method 2 to use npm directly instead of using the make file. It requires less dependencies and can easily be run from any OS.

tharvik replied:
ho right, that's way more compatible!

cewert commented on DEVGUIDE.md:
````diff
@@ -147,7 +151,7 @@ git pull
+make build-dev install
 ```
 
 ## Command Line Workflow
````
This whole section can be removed or merged with method 2/3. VSCode has a debugger and automatically connects to telnet and displays logs when deploying the app.

cewert commented on DEVGUIDE.md:
````diff
@@ -156,18 +160,12 @@ Modify code -> `make install` -> Use Roku remote to test changes -> `telnet ${RO
 
 ```bash
-make prep_commit
+make format lint
````
What's the benefit of running these through the make file? I think we should use npm or at least list instructions for both.

We can combine all the npm scripts into one "prep-commit" command if you think that'd be better. I know there are a lot of lint commands.

tharvik replied:
yep, that's a good idea; changing `npm run lint` to run all of theses.

cewert commented on DEVGUIDE.md:
```diff
@@ -43,7 +43,17 @@ Open up the new folder:
-Install Dependencies:
+## Install Dependencies
+
+You'll need [`npm`](https://github.com/npm/cli), [`make`](https://www.gnu.org/software/make) and [`curl`](https://curl.se).
```
npm is needed for all methods so it makes sense here but the others are only needed for method 2 and 3(although I think method 2 should use npm or at least list instructions for both).

The link for npm depends on OS, yea? For windows at least, npm is bundled with node js so it would make more sense to link to the nodejs install.

tharvik replied:
> The link for npm depends on OS, yea? For windows at least, npm is bundled with node js so it would make more sense to link to the nodejs install.

I think that any OS can install it from this repo, but having them directly install node is way simpler, updating it to https://nodejs.org

tharvik commented:
so I reworked it more, maybe its overeaching, tell me what you think
* we now have only two methods, one VScode, one CLI, the latter using `make`. a single tool is easier than mixing `npm` and `make` IMO and some calls can only be made via `make` (`install`, key presses, `screenshot`, …)
* "Deploy" and "Bug/Crash Reports" has been moved to "Method 2" as theses were CLI specific
  * if you want to had some precision for the VScode part, please do, I don't know the first thing about it
* "Committing" is kept at the ~top level and using `npm` as you suggested
* "Update Images" reference "Method 2" but kept as top level, as I don't think there is a way to do it via VScode (maybe?)
so now, there is a clearer separation of how to dev with each tool: debugging via VScode or use the CLI and telnet; deploying? press run on VScode or run `make install` on CLI

> changing `npm run lint` to run all of theses.

I didn't touched the .github/workflows/lint.yml but it can be easily rewritten to just run `lint` now (excepted for xml checks maybe?).

> > while at it, for `bsconfig-tests.json`, why does it actively doesn't create a package? running the suite via `make build-tests install` doesn't work as I wanted to
> 
> You can change it to make a zip by setting the `createPackage` var to true [here](https://github.com/jellyfin/jellyfin-roku/blob/unstable/bsconfig-tests.json#L29).

perfect! added a commit to generate the pkg (by removing the line thus using the default as in the other bsconfig) :)

tharvik commented:
btw, I'm using [`git commit --fixup`](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---fixupamendrewordltcommitgt) to allow for easier rebasing w/ autosquash, to not discard the comments and make a nice history.

cewert commented:
Sorry I created a couple merge conflicts. If you have any questions or you just want me to do it give me a ping.

tharvik commented:
> Sorry I created a couple merge conflicts.

no worries, it was easy :)

cewert commented:
> so I reworked it more, maybe its overeaching, tell me what you think
> 
>     * we now have only two methods, one VScode, one CLI, the latter using `make`. a single tool is easier than mixing `npm` and `make` IMO and some calls can only be made via `make` (`install`, key presses, `screenshot`, …)
> 
>     * "Deploy" and "Bug/Crash Reports" has been moved to "Method 2" as theses were CLI specific
>       
>       * if you want to had some precision for the VScode part, please do, I don't know the first thing about it
> 
>     * "Committing" is kept at the ~top level and using `npm` as you suggested
> 
>     * "Update Images" reference "Method 2" but kept as top level, as I don't think there is a way to do it via VScode (maybe?)
> 

I think you did a great job 👍  I was a little bummed at first to see method 2 get cut but it's essentially the same thing as the `Beta Test` instructions on the readme and since the file is called DEVGUIDE and is meant to be more advanced deployment methods, I think it makes perfect sense to remove it.

`Update Images` hasn't been needed in years but maybe it should be under Method 2 since thats the only method that supports it? I don't think it's a big deal either way just seems odd to have everything organized except that one thing.

> 
> > changing `npm run lint` to run all of theses.
> 
> I didn't touched the .github/workflows/lint.yml but it can be easily rewritten to just run `lint` now (excepted for xml checks maybe?).

I'm glad you didn't 😄 Technically we could change the workflows to use `npm run lint` but the overall UX would suffer. You have all the commands chained together with && (which is fine) but if the first command fails then the rest aren't run. So if your code made every lint command fail, you wouldn't know it or be able to fix them all. You'd have to fix the first error, rerun lint, fix error, rerun lint etc. Also, if they were all together, you wouldn't be able to see which lint command produced the error from the PR GUI. You'd have to open up the job log to see which command produced the error.

We will have to update the workflow though since you changed the command that lints brightscript.

There was a couple little things I saw in the devguide I'll leave comments for soon

cewert commented on docs/DEVGUIDE.md:
````diff
@@ -43,7 +41,17 @@ Open up the new folder:
+sudo apt-get install npm
+```
+
+Then, use `npm` to install more dependencies
````
```suggestion
Then, use `npm` to install dependencies
```

More doesn't sound right to me. `app dependencies` or `required dependencies` would also work

cewert commented on docs/DEVGUIDE.md:
```diff
@@ -126,57 +133,39 @@ make install
 ## (Optional) Update Images
 
-This repo already contains all necessary images for the app. This script only needs to be run when the [official Jellyfin images](https://github.com/jellyfin/jellyfin-ux) are changed to allow us to update the repo images.
+This repo already contains all necessary images for the app. This script only needs to be run when the [official Jellyfin images](https://github.com/jellyfin/jellyfin-ux) are changed to allow us to update the repo images. It only works using the [Command Line Method](method-2-command-line).
```
```suggestion
> NOTE: This script is only compatible with [Method 2: Command Line](method-2-command-line).
This repo already contains all necessary images for the app. This script only needs to be run when the [official Jellyfin images](https://github.com/jellyfin/jellyfin-ux) are changed to allow us to update the repo images.
```

If you leave it as a top level, I think this is more clear.

tharvik replied:
moved to Method 2 block and removed the line referencing the method.

cewert commented on package.json:
```diff
@@ -34,7 +34,8 @@
     "docs": "npx rimraf docs/api && jsdoc -c jsdoc.json -d docs/api",
     "format": "npx bsfmt --write",
-    "lint": "bslint",
+    "lint": "npm run check-formatting && npm run lint-bs && npm run lint-json && npm run lint-markdown && npm run lint-spelling && npm run validate",
```
I think the order could be improved. Since one command throwing an error will cause the whole chain of commands to exit, I think it would make sense to have the list of commands in order of priority.

Would basically just need to move validate from the far right to right after lint-bs maybe? Validate will tell you if the build fails

cewert replied:
I mentioned it in a comment earlier that we'd need to update all workflows that currently uses `npm run lint` but it looks like there aren't any. the brightscript lint job is using `npm run validate` instead.

tharvik replied:
moved `validate` right after `lint-bs`.

yeah, that's a shame that linting stops at the first failure, we could have a real script taking care of running each and failing only afterward, but I think we would be better off moving to real linting organiser such as [pre-commit](https://pre-commit.com) instead of redoing it ourself.

cewert replied:
> yeah, that's a shame that linting stops at the first failure, we could have a real script taking care of running each and failing only afterward, but I think we would be better off moving to real linting organiser such as [pre-commit](https://pre-commit.com) instead of redoing it ourself.

I've never heard of something like that. Thanks for telling me about it. There's also an [npm version](https://www.npmjs.com/package/pre-commit) we could use.

cewert commented on docs/DEVGUIDE.md:
```diff
@@ -43,7 +41,17 @@ Open up the new folder:
+
+You'll need [`npm`](https://nodejs.org), the default package manager for Node.js.
+
+For Debian or Ubuntu, install it via:
```
```suggestion
To install on Debian based distros:
```

Maybe this is nitpicking but if feels weird to call out these two distros and not Mint or any other debian based distros.

Another solution would be to change the text to read `To install on Linux:` and then have a line for each major package manager but this might be overkill and I'm not even sure if there's an npm package available for each of them.
```bash
sudo apt-get install npm # Debian based distros

```

cewert reviewed (changes requested):
A few small things. I still need to test this 😄 Thank you for your patience.

tharvik commented:
> I tried to test this using WSL with ubuntu 22.04 but that didn't work because `sudo apt install npm` gave me nodejs version 10

hum, [ubuntu 22.04 nodejs pkg](https://packages.ubuntu.com/jammy/nodejs) is version 12, maybe you simply need to `apt update && apt full-upgrade`?

cewert commented:
I had originally installed ubuntu 22 using chocolaty (or some other windows package manager) so apparently that was my issue because after I reinstalled ubuntu 22 using the windows store I got nodejs v12 when i installed npm.

That being said, is npm the right package to install? After installing npm I ran `npm install` in WSL and I got a bunch of warnings about dependencies needing a higher version of node:
```
charlie@DESKTOP-23CQ5C2:/mnt/c/Users/Charlie/Documents/PROJECTS/tharvik$ npm install
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'commander@10.0.1',
npm WARN EBADENGINE   required: { node: '>=14' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'coveralls-next@4.2.0',
npm WARN EBADENGINE   required: { node: '>=14' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'foreground-child@3.1.1',
npm WARN EBADENGINE   required: { node: '>=14' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'html-minifier-terser@7.2.0',
npm WARN EBADENGINE   required: { node: '^14.13.1 || >=16.0.0' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'jackspeak@2.3.6',
npm WARN EBADENGINE   required: { node: '>=14' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'markdownlint@0.31.1',
npm WARN EBADENGINE   required: { node: '>=16' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'markdownlint-cli2@0.10.0',
npm WARN EBADENGINE   required: { node: '>=16' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'strip-json-comments@5.0.1',
npm WARN EBADENGINE   required: { node: '>=14.16' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'markdownlint-micromark@0.1.7',
npm WARN EBADENGINE   required: { node: '>=16' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'minipass@7.0.4',
npm WARN EBADENGINE   required: { node: '>=16 || 14 >=14.17' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'path-scurry@1.10.1',
npm WARN EBADENGINE   required: { node: '>=16 || 14 >=14.17' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'lru-cache@10.0.1',
npm WARN EBADENGINE   required: { node: '14 || >=16.14' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'rimraf@5.0.5',
npm WARN EBADENGINE   required: { node: '>=14' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'glob@10.3.10',
npm WARN EBADENGINE   required: { node: '>=16 || 14 >=14.17' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'minimatch@9.0.3',
npm WARN EBADENGINE   required: { node: '>=16 || 14 >=14.17' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'signal-exit@4.1.0',
npm WARN EBADENGINE   required: { node: '>=14' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'yaml@2.3.2',
npm WARN EBADENGINE   required: { node: '>= 14' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }

> jellyfin-roku@1.6.6 postinstall
> npm run ropm
...
```

I know it's possible for me to manually install a higher version but shouldn't the docs give instructions to install a higher version of node? Looks like we'd need at least `v14.17`.

cewert commented:
The latest LTS [releases for nodejs](https://nodejs.dev/en/about/releases/) that are still receiving updates are 18 and 20. Support for 16 just ended a month ago. If we end up copying all the instructions for debian we might as well recommend one of the newer versions. Otherwise we'll just run into the same problem with dependencies when they eventually stop supporting v14

That's crazy that the 22.04 LTS version of ubuntu would choose to use nodejs v12. Ubuntu 22.04 came out April 21st 2022 and is supported for 3 years while nodejs v12 stopped receiving updates 9 days later on April 30th 2022 🤔

tharvik commented:
> That's crazy that the 22.04 LTS version of ubuntu would choose to use nodejs v12. Ubuntu 22.04 came out April 21st 2022 and is supported for 5 years while nodejs v12 stopped receiving updates 9 days later on April 30th 2022 🤔

yeah, the Ubuntu LTS are not really following upstream's LTS as their developpement cycles and length of LTS doesn't always match. they even ended up supporting a non-LTS kernel some years ago: that was cumbersome to say the least.

> The latest LTS [releases for nodejs](https://nodejs.dev/en/about/releases/) that are still receiving updates are 18 and 20. Support for 16 just ended a month ago. If we end up copying all the instructions for debian we might as well recommend one of the newer versions. Otherwise we'll just run into the same problem with dependencies when they eventually stop supporting v14

I'm afraid we are falling into the system-specific rabbit hole now: there is two package manager on Ubuntu now, `apt` and `snap`, the first is slowing getting dropped there but not on the Debian side. I can add instruction on how to add deps via `apt` (mostly a copypaste of https://deb.nodesource.com) which fits both for now but it will probably drift slowly. IMO, we can simply state that one need to have nodejs >= 16, get rid of the debian-based blocks and that should do the trick; what do you prefer?

cewert commented:
> IMO, we can simply state that one need to have nodejs >= 16, get rid of the debian-based blocks and that should do the trick; what do you prefer?

That works for me 👍

cewert commented:
The docs say to run `make build-dev`, then if you want to install it on the device to run `make install` but `make install` is also running `make build-dev` so we end up building twice. Seems like we could use a `make deploy` that only deploys the app without building it, yea?

I'm not sure where this would fit best, but it would be nice to have a list of supported make commands somewhere to have as a quick reference. 

PS: I'm still using WSL to test. I don't know why but using WSL makes the npm run commands take **forever**. `make lint` took somewhere between 10 and 20 minutes. I don't exactly know because I had to go afk it took so long.

cewert commented on Makefile:
```diff
@@ -1,28 +1,93 @@
+
+## development
+
+node_modules/: package-lock.json; npm ci
```
Why are we running `npm ci` so much? It only needs to be run once but is currently being run for all of these commands:
- all 3 builds
- format
- lint

tharvik replied:
hum, I'm not sure I understand you: it is only ran once, but depended by multiple targets. make is a file-based build system, not a task one.
in this case, "node_modules/" is created by running `npm ci`, and all target using `npm run ...` depends on it. when a target, let's say "build-dev" is asked to be built, `make` will first check if "node_modules" exists, if it doesn't it will get built, if it does, it will skip it.
the `.PHONY` target is a special one that states that instead of checking deps, `make` should always run the command. it is the dedicated way to using `make` as a task build-system.

cewert replied:
Makes sense. Thanks for the explanation.

cewert commented on Makefile:
```diff
@@ -1,28 +1,93 @@
+build-tests: node_modules/; npm run build-tests
+
+# default to build-dev if file doesn't exist
+out/jellyfin-roku.zip:; $(MAKE) build-dev
```
When using `npm run build`, the zip file is named after the folder i.e. `./out/folder-name.zip`.

tharvik replied:
ho, didn't knew that, fixed in rebase.

cewert commented on Makefile:
```diff
@@ -1,28 +1,93 @@
+EXTRACT_ERROR_CMD := grep "<font color" | sed "s/<font color=\"red\">//" | sed "s[</font>[["
+.PHONY: install remove
+install: out/jellyfin-roku.zip home
+	$(CURL_LOGGED_CMD) -F "mysubmit=Install" -F "archive=@$<" -F "passwd=" http://$(ROKU_DEV_TARGET)/plugin_install | $(EXTRACT_ERROR_CMD)
```
I couldn't get `make install` to deploy the app on my roku.

```
make[1]: Leaving directory '/mnt/c/Users/Charlie/Documents/PROJECTS/tharvik'
curl --show-error -XPOST http://192.168.1.177:8060/keypress/home
sleep 2 # wait for roku reaction
curl --show-error --user rokudev:pw--digest -F "mysubmit=Install" -F "archive=@out/jellyfin-roku.zip" -F "passwd=" http://192.168.1.177/plugin_install | grep "<font color" | sed "s/<font color=\"red\">//" | sed "s[</font>[["
curl: (26) Failed to open/read local data from file/application
```

cewert replied:
This was probably because I wasn't using the default folder name. I cloned the repo then renamed the folder to `tharvik`

tharvik replied:
indeed, `curl` tries to read a non-existing file, should be fixed by the next rebase.

tharvik commented:
> The docs say to run `make build-dev`, then if you want to install it on the device to run `make install` but `make install` is also running `make build-dev` so we end up building twice.

it doesn't build it twice on my system (Linux w/ GNU Make 4.4.1), neither should it. the package is build only if it doesn't exists already (defaulting to build-dev in this case). that allows for users to build-{dev,prod,tests} at their will.

I know of some issues with old FAT filesystem (it does have timestamp with a 2-second resolution which might confuse it on rapidly firing make), but I don't think anyone is building code directly on it.

> I'm not sure where this would fit best, but it would be nice to have a list of supported make commands somewhere to have as a quick reference.

adding a `help` target listing theses.

> This pull request has merge conflicts. Please resolve the conflicts so the PR can be reviewed. Thanks!

and there goes the rebase.

cewert commented:
Can we show a more helpful error message when the env vars are missing? This is the error I got when running `make install` without setting my env vars first:

```
charlie@DESKTOP-23CQ5C2:/mnt/c/Users/Charlie/Documents/PROJECTS/tharvik$ make install
make: *** No rule to make target 'install'.  Stop.
```

Also, can we show the full list of commands when running `make help`? I think it's clever to tell the user which commands are linked to the env vars but I still think we should display them all.

When running `make install` can we still launch the app even if it's identical? `make install` normally launches the app except when deploying an identical package
```
charlie@DESKTOP-23CQ5C2:/mnt/c/Users/Charlie/Documents/PROJECTS/tharvik$ make install
curl --show-error -XPOST http://192.168.1.177:8060/keypress/home
sleep 2 # wait for roku reaction
curl --show-error --user rokudev:pw --digest -F "mysubmit=Install" -F "archive=@out/tharvik.zip" -F "passwd=" http://192.168.1.177/plugin_install | grep "<font color" | sed "s/<font color=\"red\">//" | sed "s[</font>[["
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0 2184k    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 2189k    0  4939  100 2184k   5415  2395k --:--:-- --:--:-- --:--:-- 1461k
    Application Received: Identical to previous version -- not replacing.
```

tharvik commented:
> Can we show a more helpful error message when the env vars are missing?

we could but that makes the Makefile way less readable as we would need to redefine the targets in every `ifdef` path combination. having the "help" target looks enough for me. IMO one would run it when something doesn't work as expected.

> Also, can we show the full list of commands when running `make help`?

updated to show each target ~group, and document the needed environment, unconditionnaly.

> When running `make install` can we still launch the app even if it's identical? `make install` normally launches the app except when deploying an identical package

added a "launch" target to do that, and running it at the end of "install".
